src/components/AST/Viz.js

```diff
@@ -0,0 +1,192 @@
+    return Object.entries(obj).map(([i, v], index) => {
+      return (
+        <Fragment key={index}>
+          {typeof v === "object" && v !== null ? (
+            <Composite k={i} ast={v} cursor={cursor} setPos={setPos} />
+          ) : (
+            <Primitive k={i} val={v} />
+          )}
+        </Fragment>
+      );
+    });
+  }, [obj, cursor, setPos]);
+
+  const { type, loc, value, name } = obj;
+  const label = type ? type : k;
+  const keys = Object.keys(obj);
+  const [active, setActive] = useState(false);
+  const highlight = useMemo(() => {
+    return (value !== undefined || name !== undefined || !active) &&
+      isBound(cursor, loc)
+      ? "highlight"
+      : "";
+  }, [active, cursor, loc, value, name]);
+
+  const popup = lookUpNodeType(type);
+  const title = `${label} : { ${keys.join(", ")} } ${highlight}`;
+  return keys.length ? (
+    <Accordion.Accordion
+      onTitleClick={() => {
+        loc &&
+          setPos({
+            anchor: fixLoc(loc).end,
+            head: fixLoc(loc).start,
+          });
+        setActive(active => !active);
+      }}
+      panels={[
+        {
+          key: 0,
+          title: popup.length ? (
+            <Accordion.Title
+              content={
+                <Popup
+                  header={type}
+                  content={<UglyPopup def={popup} />}
+                  on={"hover"}
+                  trigger={<span>{title}</span>}
+                />
+              }
+            />
+          ) : (
+            title
+          ),
+          content: content,
+        },
+      ]}
+    />
+  ) : (
+    <Primitive k={label} val={"{ }"} />
+  );
+}
+
+function CompositeArr({ k, arr, cursor, setPos }) {
+  // setPos is memoized so dependency is just to placate warnings.
+  const components = useMemo(() => {
+    return arr.map((value, index) => {
+      return (
+        <Fragment key={index}>
+          {typeof value === "object" && value !== null ? (
+            <Composite k={k} ast={value} cursor={cursor} setPos={setPos} />
+          ) : (
+            <Primitive k={k} ast={value} />
+          )}
+        </Fragment>
+      );
+    });
+  }, [k, arr, cursor, setPos]);
+
+  const [active, setActive] = useState(false);
+  const highlight = useMemo(() => {
+    return !active && arr.some(x => isBound(cursor, x.loc)) ? "highlight" : "";
+  }, [active, arr, cursor]);
+
+  return arr.length ? (
+    <Accordion.Accordion
+      onTitleClick={() => {
+        setActive(active => !active);
+      }}
+      panels={[
+        {
+          key: 0,
+          title: `${k} : [ ${arr.length} element ] ${highlight}`,
+          content: components,
+        },
+      ]}
+    />
+  ) : (
+    <Primitive k={k} val={"[ ]"} />
+  );
+}
+
+function Composite({ k, ast, cursor, setPos }) {
+  return !Array.isArray(ast) ? (
+    <CompositeObj k={k} obj={ast} cursor={cursor} setPos={setPos} />
+  ) : (
+    <CompositeArr k={k} arr={ast} cursor={cursor} setPos={setPos} />
+  );
+}
```

Comment:
I think it'll be a good idea to add some quick comments explaining the difference between `CompositeObj` and `CompositeArr` since the code is similar - just to avoid confusion in the long run for future contributors.

Reply:
Done!